File: backend/internal/models/community.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// ModerationStatus represents the moderation state of a community post
type ModerationStatus string

const (
	ModerationStatusPending  ModerationStatus = "pending"
	ModerationStatusApproved ModerationStatus = "approved"
	ModerationStatusRejected ModerationStatus = "rejected"
)

// PostTagSource indicates where the tagged product came from
type PostTagSource string

const (
	PostTagSourceProduct PostTagSource = "product" // Direct product selection from catalog
	PostTagSourceOrder   PostTagSource = "order"   // Product selected from user's order history
)

// ProductCustomizationTag represents customization options selected for a tagged product
type ProductCustomizationTag struct {
	SelectedMetal   string            `json:"selectedMetal,omitempty" bson:"selectedMetal,omitempty"`
	SelectedPlating string            `json:"selectedPlating,omitempty" bson:"selectedPlating,omitempty"`
	StoneColors     map[string]string `json:"stoneColors,omitempty" bson:"stoneColors,omitempty"`
	SelectedSize    string            `json:"selectedSize,omitempty" bson:"selectedSize,omitempty"`
	EngravingText   string            `json:"engravingText,omitempty" bson:"engravingText,omitempty"`
	Thickness       *float64          `json:"thickness,omitempty" bson:"thickness,omitempty"`
}

// ProductTag represents a product tagged in a community post
type ProductTag struct {
	ID            string                   `json:"id" bson:"id"`
	Name          string                   `json:"name" bson:"name"`
	Price         float64                  `json:"price" bson:"price"`
	ImageURL      string                   `json:"imageUrl" bson:"imageUrl"`
	Customization *ProductCustomizationTag `json:"customization,omitempty" bson:"customization,omitempty"`
	OrderID       string                   `json:"orderId,omitempty" bson:"orderId,omitempty"`
	OrderNumber   string                   `json:"orderNumber,omitempty" bson:"orderNumber,omitempty"`
}

// OrderTag represents an order tagged in a community post
type OrderTag struct {
	OrderID     string       `json:"orderId" bson:"orderId"`
	OrderNumber string       `json:"orderNumber" bson:"orderNumber"`
	OrderDate   time.Time    `json:"orderDate" bson:"orderDate"`
	OrderTotal  float64      `json:"orderTotal" bson:"orderTotal"`
	Products    []ProductTag `json:"products" bson:"products"`
}

// CommunityPost represents a post in the community section
type CommunityPost struct {
	ID               primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	UserID           primitive.ObjectID `json:"userId" bson:"userId" validate:"required"`
	UserName         string             `json:"userName" bson:"userName"`
	UserAvatar       string             `json:"userAvatar,omitempty" bson:"userAvatar,omitempty"`
	Content          string             `json:"content" bson:"content" validate:"required,min=1,max=5000"`
	Images           []string           `json:"images,omitempty" bson:"images,omitempty"`
	Videos           []string           `json:"videos,omitempty" bson:"videos,omitempty"`
	LikeCount        int                `json:"likeCount" bson:"likeCount"`
	VoteCount        int                `json:"voteCount" bson:"voteCount"`
	CommentCount     int                `json:"commentCount" bson:"commentCount"`
	Tags             []string           `json:"tags,omitempty" bson:"tags,omitempty"`
	// Product/Order tagging
	Products         []ProductTag       `json:"products,omitempty" bson:"products,omitempty"`
	Order            *OrderTag          `json:"order,omitempty" bson:"order,omitempty"`
	TagSource        PostTagSource      `json:"tagSource,omitempty" bson:"tagSource,omitempty"`
	IsAdminPost      bool               `json:"isAdminPost" bson:"isAdminPost"`
	IsFeatured       bool               `json:"isFeatured" bson:"isFeatured"`
	IsPinned         bool               `json:"isPinned" bson:"isPinned"`
	// Moderation fields
	ModerationStatus ModerationStatus   `json:"moderationStatus" bson:"moderationStatus"`
	RejectionReason  string             `json:"rejectionReason,omitempty" bson:"rejectionReason,omitempty"`
	ModeratedBy      primitive.ObjectID `json:"moderatedBy,omitempty" bson:"moderatedBy,omitempty"`
	ModeratedAt      *time.Time         `json:"moderatedAt,omitempty" bson:"moderatedAt,omitempty"`
	CreatedAt        time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt        time.Time          `json:"updatedAt" bson:"updatedAt"`
}

// PostLike represents a like on a community post
type PostLike struct {
	ID        primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	PostID    primitive.ObjectID `json:"postId" bson:"postId" validate:"required"`
	UserID    primitive.ObjectID `json:"userId" bson:"userId" validate:"required"`
	UserName  string            `json:"userName" bson:"userName"`
	CreatedAt time.Time         `json:"createdAt" bson:"createdAt"`
}

// PostVote represents a vote on a community post
type PostVote struct {
	ID        primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	PostID    primitive.ObjectID `json:"postId" bson:"postId" validate:"required"`
	UserID    primitive.ObjectID `json:"userId" bson:"userId" validate:"required"`
	UserName  string            `json:"userName" bson:"userName"`
	VoteType  string            `json:"voteType" bson:"voteType" validate:"required,oneof=up down"` // up or down
	CreatedAt time.Time         `json:"createdAt" bson:"createdAt"`
}

// PostComment represents a comment on a community post
type PostComment struct {
	ID        primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	PostID    primitive.ObjectID `json:"postId" bson:"postId" validate:"required"`
	UserID    primitive.ObjectID `json:"userId" bson:"userId" validate:"required"`
	UserName  string            `json:"userName" bson:"userName"`
	UserAvatar string           `json:"userAvatar,omitempty" bson:"userAvatar,omitempty"`
	Content   string            `json:"content" bson:"content" validate:"required,min=1,max=1000"`
	LikeCount int               `json:"likeCount" bson:"likeCount"`
	CreatedAt time.Time         `json:"createdAt" bson:"createdAt"`
	UpdatedAt time.Time         `json:"updatedAt" bson:"updatedAt"`
}

// InstagramProfile represents a user's Instagram integration
type InstagramProfile struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	UserID        primitive.ObjectID `json:"userId" bson:"userId" validate:"required"`
	InstagramID   string            `json:"instagramId" bson:"instagramId" validate:"required"`
	Username      string            `json:"username" bson:"username" validate:"required"`
	DisplayName   string            `json:"displayName,omitempty" bson:"displayName,omitempty"`
	ProfilePicURL string            `json:"profilePicUrl,omitempty" bson:"profilePicUrl,omitempty"`
	Bio           string            `json:"bio,omitempty" bson:"bio,omitempty"`
	FollowerCount int               `json:"followerCount,omitempty" bson:"followerCount,omitempty"`
	FollowingCount int              `json:"followingCount,omitempty" bson:"followingCount,omitempty"`
	PostCount     int               `json:"postCount,omitempty" bson:"postCount,omitempty"`
	IsVerified    bool              `json:"isVerified" bson:"isVerified"`
	IsActive      bool              `json:"isActive" bson:"isActive"`
	LinkedAt      time.Time         `json:"linkedAt" bson:"linkedAt"`
	UpdatedAt     time.Time         `json:"updatedAt" bson:"updatedAt"`
}

// CreatePostRequest represents the request to create a community post
type CreatePostRequest struct {
	Content    string        `json:"content" validate:"required,min=1,max=5000"`
	Images     []string      `json:"images,omitempty"`
	Videos     []string      `json:"videos,omitempty"`
	Tags       []string      `json:"tags,omitempty"`
	Products   []ProductTag  `json:"products,omitempty"`   // Tagged products
	Order      *OrderTag     `json:"order,omitempty"`      // Tagged order
	TagSource  PostTagSource `json:"tagSource,omitempty"`  // Source of tagged item
	IsFeatured bool          `json:"isFeatured,omitempty"`
	IsPinned   bool          `json:"isPinned,omitempty"`
}

// UpdatePostRequest represents the request to update a community post
type UpdatePostRequest struct {
	Content    *string  `json:"content,omitempty" validate:"omitempty,min=1,max=5000"`
	Images     []string `json:"images,omitempty"`
	Videos     []string `json:"videos,omitempty"`
	Tags       []string `json:"tags,omitempty"`
	IsFeatured *bool    `json:"isFeatured,omitempty"`
	IsPinned   *bool    `json:"isPinned,omitempty"`
}

// CreateCommentRequest represents the request to create a comment
type CreateCommentRequest struct {
	PostID  primitive.ObjectID `json:"postId" validate:"required"`
	Content string            `json:"content" validate:"required,min=1,max=1000"`
}

// LinkInstagramRequest represents the request to link Instagram profile
type LinkInstagramRequest struct {
	InstagramID   string `json:"instagramId" validate:"required"`
	Username      string `json:"username" validate:"required"`
	DisplayName   string `json:"displayName,omitempty"`
	ProfilePicURL string `json:"profilePicUrl,omitempty"`
	Bio           string `json:"bio,omitempty"`
}

// ModeratePostRequest represents the request to moderate a community post
type ModeratePostRequest struct {
	Action string `json:"action" validate:"required,oneof=approve reject"` // approve or reject
	Reason string `json:"reason,omitempty"`                                 // Required when rejecting
}

// CommunityFeedResponse represents the response for community feed
type CommunityFeedResponse struct {
	Posts      []CommunityPost `json:"posts"`
	Total      int64          `json:"total"`
	Page       int            `json:"page"`
	Limit      int            `json:"limit"`
	TotalPages int            `json:"totalPages"`
}

// PostEngagement represents engagement stats for a post
type PostEngagement struct {
	PostID       primitive.ObjectID `json:"postId"`
	LikeCount    int               `json:"likeCount"`
	VoteCount    int               `json:"voteCount"`
	CommentCount int               `json:"commentCount"`
	UserLiked    bool              `json:"userLiked"`
	UserVoted    string            `json:"userVoted,omitempty"` // "up", "down", or empty
}

File: lib/models/community.dart
class CommunityPost {
  final String id;
  final String userId;
  final String userName;
  final String? userAvatar;
  final String content;
  final List<String> images;
  final List<String> videos;
  final int likeCount;
  final int voteCount;
  final int commentCount;
  final List<String> tags;
  final bool isAdminPost;
  final bool isFeatured;
  final bool isPinned;
  final DateTime createdAt;
  final DateTime updatedAt;

  CommunityPost({
    required this.id,
    required this.userId,
    required this.userName,
    this.userAvatar,
    required this.content,
    this.images = const [],
    this.videos = const [],
    this.likeCount = 0,
    this.voteCount = 0,
    this.commentCount = 0,
    this.tags = const [],
    this.isAdminPost = false,
    this.isFeatured = false,
    this.isPinned = false,
    required this.createdAt,
    required this.updatedAt,
  });

  factory CommunityPost.fromJson(Map<String, dynamic> json) {
    return CommunityPost(
      id: json['id']?.toString() ?? '',
      userId: json['userId']?.toString() ?? '',
      userName: json['userName']?.toString() ?? '',
      userAvatar: json['userAvatar']?.toString(),
      content: json['content']?.toString() ?? '',
      images: List<String>.from(json['images'] ?? []),
      videos: List<String>.from(json['videos'] ?? []),
      likeCount: json['likeCount'] ?? 0,
      voteCount: json['voteCount'] ?? 0,
      commentCount: json['commentCount'] ?? 0,
      tags: List<String>.from(json['tags'] ?? []),
      isAdminPost: json['isAdminPost'] ?? false,
      isFeatured: json['isFeatured'] ?? false,
      isPinned: json['isPinned'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'userName': userName,
      'userAvatar': userAvatar,
      'content': content,
      'images': images,
      'videos': videos,
      'likeCount': likeCount,
      'voteCount': voteCount,
      'commentCount': commentCount,
      'tags': tags,
      'isAdminPost': isAdminPost,
      'isFeatured': isFeatured,
      'isPinned': isPinned,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class PostComment {
  final String id;
  final String postId;
  final String userId;
  final String userName;
  final String? userAvatar;
  final String content;
  final int likeCount;
  final DateTime createdAt;
  final DateTime updatedAt;

  PostComment({
    required this.id,
    required this.postId,
    required this.userId,
    required this.userName,
    this.userAvatar,
    required this.content,
    this.likeCount = 0,
    required this.createdAt,
    required this.updatedAt,
  });

  factory PostComment.fromJson(Map<String, dynamic> json) {
    return PostComment(
      id: json['id']?.toString() ?? '',
      postId: json['postId']?.toString() ?? '',
      userId: json['userId']?.toString() ?? '',
      userName: json['userName']?.toString() ?? '',
      userAvatar: json['userAvatar']?.toString(),
      content: json['content']?.toString() ?? '',
      likeCount: json['likeCount'] ?? 0,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'postId': postId,
      'userId': userId,
      'userName': userName,
      'userAvatar': userAvatar,
      'content': content,
      'likeCount': likeCount,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class PostEngagement {
  final String postId;
  final int likeCount;
  final int voteCount;
  final int commentCount;
  final bool userLiked;
  final String userVoted; // "up", "down", or ""

  PostEngagement({
    required this.postId,
    this.likeCount = 0,
    this.voteCount = 0,
    this.commentCount = 0,
    this.userLiked = false,
    this.userVoted = '',
  });

  factory PostEngagement.fromJson(Map<String, dynamic> json) {
    return PostEngagement(
      postId: json['postId']?.toString() ?? '',
      likeCount: json['likeCount'] ?? 0,
      voteCount: json['voteCount'] ?? 0,
      commentCount: json['commentCount'] ?? 0,
      userLiked: json['userLiked'] ?? false,
      userVoted: json['userVoted']?.toString() ?? '',
    );
  }
}

class InstagramProfile {
  final String id;
  final String userId;
  final String instagramId;
  final String username;
  final String? displayName;
  final String? profilePicUrl;
  final String? bio;
  final int followerCount;
  final int followingCount;
  final int postCount;
  final bool isVerified;
  final bool isActive;
  final DateTime linkedAt;
  final DateTime updatedAt;

  InstagramProfile({
    required this.id,
    required this.userId,
    required this.instagramId,
    required this.username,
    this.displayName,
    this.profilePicUrl,
    this.bio,
    this.followerCount = 0,
    this.followingCount = 0,
    this.postCount = 0,
    this.isVerified = false,
    this.isActive = true,
    required this.linkedAt,
    required this.updatedAt,
  });

  factory InstagramProfile.fromJson(Map<String, dynamic> json) {
    return InstagramProfile(
      id: json['id']?.toString() ?? '',
      userId: json['userId']?.toString() ?? '',
      instagramId: json['instagramId']?.toString() ?? '',
      username: json['username']?.toString() ?? '',
      displayName: json['displayName']?.toString(),
      profilePicUrl: json['profilePicUrl']?.toString(),
      bio: json['bio']?.toString(),
      followerCount: json['followerCount'] ?? 0,
      followingCount: json['followingCount'] ?? 0,
      postCount: json['postCount'] ?? 0,
      isVerified: json['isVerified'] ?? false,
      isActive: json['isActive'] ?? true,
      linkedAt: json['linkedAt'] != null
          ? DateTime.parse(json['linkedAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'instagramId': instagramId,
      'username': username,
      'displayName': displayName,
      'profilePicUrl': profilePicUrl,
      'bio': bio,
      'followerCount': followerCount,
      'followingCount': followingCount,
      'postCount': postCount,
      'isVerified': isVerified,
      'isActive': isActive,
      'linkedAt': linkedAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class CommunityFeedResponse {
  final List<CommunityPost> posts;
  final int total;
  final int page;
  final int limit;
  final int totalPages;

  CommunityFeedResponse({
    required this.posts,
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
  });

  factory CommunityFeedResponse.fromJson(Map<String, dynamic> json) {
    return CommunityFeedResponse(
      posts: (json['posts'] as List<dynamic>?)
              ?.map((post) => CommunityPost.fromJson(post as Map<String, dynamic>))
              .toList() ??
          [],
      total: json['total'] ?? 0,
      page: json['page'] ?? 1,
      limit: json['limit'] ?? 20,
      totalPages: json['totalPages'] ?? 1,
    );
  }
}
